File: C++/1046-shortest_distance.cc
#include <iostream>
#include <vector>


int main() {

    std::ios::sync_with_stdio(false);

    int N;
    std::cin >> N;

    std::vector<int> cum_sum = {0};
    while (N--) {
        int num;
        std::cin >> num;
        cum_sum.push_back(cum_sum.back() + num);
    }

    int M;
    std::cin >> M;
    while (M--) {
        int i, j;
        std::cin >> i >> j;
        --i, --j;
        if (j < i) {
            std::swap(i, j);
        }
        int dist = cum_sum[j] - cum_sum[i];
        int alt_dist = cum_sum[i] + cum_sum.back() - cum_sum[j];
        std::cout << (dist < alt_dist ? dist : alt_dist) << '\n';
    }

    return 0;
}

File: Python/1046-shortest_distance.py


if __name__ == '__main__':
    nums = [int(i) for i in input().split()][1:]

    cum_sum = [0]
    for num in nums:
        cum_sum.append(cum_sum[-1] + num)

    n_pair = int(input())
    for _ in range(n_pair):
        i, j = [int(i) - 1 for i in input().split()]
        if j < i:
            i, j = j, i
        dist = cum_sum[j] - cum_sum[i]
        alt_dist = cum_sum[i] + cum_sum[-1] - cum_sum[j]
        print(min(dist, alt_dist))
